File: java/src/leetcode/lists/offerII/Offer53.java
package leetcode.lists.offerII;

import common.TreeNode;

import java.util.Stack;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/7/25 17:02
 */
public class Offer53 {


    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {

        if (root == null) {
            return null;
        }
        TreeNode pre = null;
        Stack<TreeNode> stack = new Stack<>();
        while (!stack.empty() || root != null) {
            while (root != null) {
                stack.add(root);
                root = root.left;
            }
            root = stack.pop();
            if (pre == p) {
                return root;
            }
            //fix
            pre = root;
            root = root.right;
        }
        return null;
    }
}

File: go/src/leetcode/offerII/Offer53.go
package offerII

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/17 17:24
 */

func inorderSuccessor(root *TreeNode, p *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	var res *TreeNode
	flag := false
	var dfs func(root *TreeNode, p *TreeNode)
	dfs = func(root *TreeNode, p *TreeNode) {
		if root == nil {
			return
		}
		dfs(root.Left, p)
        if res != nil {
            return
        }
		if root == p {
			flag = true
		}
		if flag && root.Val > p.Val {
			res = root
			flag = false
		}
		dfs(root.Right, p)
	}
	dfs(root, p)
	return res
}
